x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmSettingsTests.java

```diff
@@ -0,0 +1,374 @@
+            JwtRealmSettings.CLAIMS_GROUPS.getClaim()
+        )) {
+            final String settingKey = RealmSettings.getFullSettingKey(REALM_NAME, setting);
+            for (final String rejectedValue : new String[] { null, "" }) {
+                final Exception exception = expectThrows(IllegalArgumentException.class, () -> {
+                    final Settings settings = Settings.builder().put(settingKey, rejectedValue).build();
+                    final String actualValue = this.buildConfig(settings).getSetting(setting);
+                    fail("No exception. Expected one for " + settingKey + "=" + rejectedValue + ". Got " + actualValue + ".");
+                });
+                assertThat(exception.getMessage(), equalTo("Invalid null or empty claim name for [" + settingKey + "]."));
+            }
+            for (final String acceptedValue : new String[] { "sub", "name", "email", "dn" }) {
+                final Settings settings = Settings.builder().put(settingKey, acceptedValue).build();
+                final String actualValue = this.buildConfig(settings).getSetting(setting);
+                assertThat(actualValue, equalTo(acceptedValue));
+            }
+        }
+    }
+
+    public void testClaimPatterns() {
+        for (final Setting.AffixSetting<String> setting : List.of(
+            JwtRealmSettings.CLAIMS_PRINCIPAL.getPattern(),
+            JwtRealmSettings.CLAIMS_GROUPS.getPattern()
+        )) {
+            final String settingKey = RealmSettings.getFullSettingKey(REALM_NAME, setting);
+            for (final String rejectedValue : new String[] { "[" }) {
+                final Exception exception = expectThrows(IllegalArgumentException.class, () -> {
+                    final Settings settings = Settings.builder().put(settingKey, rejectedValue).build();
+                    final String actualValue = this.buildConfig(settings).getSetting(setting);
+                    fail("No exception. Expected one for " + settingKey + "=" + rejectedValue + ". Got " + actualValue + ".");
+                });
+                assertThat(exception.getMessage(), equalTo("Invalid claim value regex pattern for [" + settingKey + "]."));
+            }
+            for (final String acceptedValue : new String[] { "^([^@]+)@example\\.com$", "^Group-(.+)$" }) {
+                final Settings settings = Settings.builder().put(settingKey, acceptedValue).build();
+                final String actualValue = this.buildConfig(settings).getSetting(setting);
+                assertThat(actualValue, equalTo(acceptedValue));
+            }
+        }
+    }
+
+    public void testPopulateUserMetadata() {
+        Setting.AffixSetting<Boolean> setting = JwtRealmSettings.POPULATE_USER_METADATA;
+        final String settingKey = RealmSettings.getFullSettingKey(REALM_NAME, setting);
+        for (final String rejectedValue : new String[] { "", "unknown", "t", "f", "TRUE", "FALSE", "True", "False" }) {
```

Comment:
Our settings parser will accept all kind of YAML boolean valid values: https://yaml.org/type/bool.html and map these to "true" and "false" . If anything else is provided, we would have failed in Setting.parseBoolean and we're testing this method in its own unit tests. Maybe we don't need this test here since we're not doing any additional validation for the POPULATE_USER_METADATA ?

Reply:
This unit test is only for ES settings parsing, no yaml preprocessing. The test inputs are valid.

I started with the same assumption you did. I tried partial boolean-like strings in the acceptedValue loop, but the ES boolean setting parser threw IllegalArgumentException. Root cause is **org.elasticsearch.core.Booleans.parseBoolean(String)** because it only accepts exact matches of lowercase "true" or "false".

I will add a comment to clarify this in the test. In other words, inputs into the yaml parser are not accepted, only outputs of the yaml parser are accepted by the ES boolean settings parser.